packages/contracts-bedrock/src/L1/winddown/BalanceClaimer.sol

```diff
@@ -0,0 +1,118 @@
+    }
+
+    function initialize(
+        address _ethBalanceWithdrawer,
+        address _erc20BalanceWithdrawer,
+        bytes32 _root
+    )
+        public
+        initializer
+    {
+        ethBalanceWithdrawer = IEthBalanceWithdrawer(_ethBalanceWithdrawer);
+        erc20BalanceWithdrawer = IErc20BalanceWithdrawer(_erc20BalanceWithdrawer);
+        root = _root;
+    }
+
+    /// @notice Claims the tokens for the user
+    /// @param _proof The merkle proof
+    /// @param _user The user address
+    /// @param _ethBalance The eth balance of the user
+    /// @param _erc20TokenBalances The ERC20 tokens balances of the user
+    function claim(
+        bytes32[] calldata _proof,
+        address _user,
+        uint256 _ethBalance,
+        IBalanceWithdrawer.Erc20BalanceClaim[] calldata _erc20TokenBalances
+    )
+        external
+    {
+        if (!_canClaim(_proof, _user, _ethBalance, _erc20TokenBalances)) {
+            revert IBalanceClaimer.NoBalanceToClaim();
+        }
+        claimed[_user] = true;
+        if (_erc20TokenBalances.length != 0) {
+            erc20BalanceWithdrawer.withdrawErc20Balance(_user, _erc20TokenBalances);
+        }
+        if (_ethBalance != 0) {
+            ethBalanceWithdrawer.withdrawEthBalance(_user, _ethBalance);
+        }
+        emit BalanceClaimed({ user: _user, ethBalance: _ethBalance, erc20TokenBalances: _erc20TokenBalances });
+    }
+
+    /// @notice Checks if the user can claim the tokens
+    /// @param _proof The merkle proof
+    /// @param _user The user address
+    /// @param _ethBalance The eth balance of the user
+    /// @param _erc20TokenBalances The ERC20 tokens balances of the user
+    /// @return _canClaimTokens True if the user can claim the tokens
+    function canClaim(
+        bytes32[] calldata _proof,
+        address _user,
+        uint256 _ethBalance,
+        IBalanceWithdrawer.Erc20BalanceClaim[] calldata _erc20TokenBalances
+    )
+        external
+        view
+        returns (bool _canClaimTokens)
+    {
+        _canClaimTokens = _canClaim(_proof, _user, _ethBalance, _erc20TokenBalances);
+    }
+
+    /// @notice Checks if the user can claim the tokens
+    /// @param _proof The merkle proof
+    /// @param _user The user address
+    /// @param _ethBalance The eth balance of the user
+    /// @param _erc20TokenBalances The ERC20 tokens balances of the user
+    /// @return _canClaimTokens True if the user can claim the tokens
+    function _canClaim(
+        bytes32[] calldata _proof,
+        address _user,
+        uint256 _ethBalance,
+        IBalanceWithdrawer.Erc20BalanceClaim[] calldata _erc20TokenBalances
+    )
+        internal
+        view
+        returns (bool _canClaimTokens)
+    {
+        if (claimed[_user]) return false;
+        bytes32 _leaf = keccak256(bytes.concat(keccak256(abi.encode(_user, _ethBalance, _erc20TokenBalances))));
+        _canClaimTokens = MerkleProof.verify(_proof, root, _leaf);
+    }
```

Comment:
is there a chance someone may need two claims to fully withdraw their balances? if so, the `claimed` check wouldn't allow a second claim.

Reply:
no, the claim should be just one per account, the merkle tree should include all the tokens and eth information for the user to claim